dynamic-config/config-apollo/src/main/java/io/servicecomb/config/client/ApolloClient.java

```diff
@@ -0,0 +1,162 @@
+
+  private String clusters = APOLLO_CONFIG.getServerClusters();
+
+  private String namespace = APOLLO_CONFIG.getNamespace();
+
+  private UpdateHandler updateHandler;
+
+  private static ClientPoolManager<HttpClientWithContext> clientMgr = new ClientPoolManager<>();
+
+  public ApolloClient(UpdateHandler updateHandler) {
+    this.updateHandler = updateHandler;
+  }
+
+  public void refreshApolloConfig() {
+    EXECUTOR
+        .scheduleWithFixedDelay(new ConfigRefresh(serviceUri), firstRefreshInterval, refreshInterval, TimeUnit.SECONDS);
+  }
+
+  class ConfigRefresh implements Runnable {
+    private String serviceUri;
+
+    public ConfigRefresh(String serviceUris) {
+      this.serviceUri = serviceUris;
+    }
+
+    @Override
+    public void run() {
+      try {
+        refreshConfig();
+      } catch (Exception e) {
+        LOGGER.error("client refresh thread exception", e);
+      }
+    }
+
+    public void refreshConfig() {
+      RestTemplate rest = new RestTemplate();
+      HttpHeaders headers = new HttpHeaders();
+      headers.add("Content-Type", "application/json;charset=UTF-8");
+      headers.add("Authorization", token);
+      HttpEntity<String> entity = new HttpEntity<String>(headers);
+      ResponseEntity<String> exchange = rest.exchange(composeAPI(), HttpMethod.GET, entity, String.class);
+      if (HttpResponseStatus.OK.code() == exchange.getStatusCode().value()) {
+        try {
+          Map<String, Object> body = JsonUtils.OBJ_MAPPER.readValue(exchange.getBody(),
+              new TypeReference<Map<String, Object>>() {
+              });
+          refreshConfigItems((Map<String, Object>) body.get("configurations"));
+        } catch (IOException e) {
+          e.printStackTrace();
+        }
+      }
+    }
+
+    private String composeAPI() {
+      String api = serviceUri + "/openapi/v1/envs/" + env +
+          "/apps/" + serviceName +
+          "/clusters/" + clusters +
+          "/namespaces/" + namespace +
+          "/releases/latest";
+      return api;
+    }
+
+    private void refreshConfigItems(Map<String, Object> map) {
+      compareChangedConfig(originalConfigMap, map);
+      originalConfigMap.clear();
+      originalConfigMap.putAll(map);
+    }
+
+    private void compareChangedConfig(Map<String, Object> before, Map<String, Object> after) {
+      Map<String, Object> itemsCreated = new HashMap<>();
+      Map<String, Object> itemsDeleted = new HashMap<>();
+      Map<String, Object> itemsModified = new HashMap<>();
+      if (before == null || before.isEmpty()) {
+        updateHandler.handle("create", after);
+        return;
+      }
+      if (after == null || after.isEmpty()) {
+        updateHandler.handle("delete", before);
+        return;
+      }
+      for (String itemKey : after.keySet()) {
+        if (!before.containsKey(itemKey)) {
+          itemsCreated.put(itemKey, after.get(itemKey));
+        } else if (!after.get(itemKey).equals(before.get(itemKey))) {
```

Comment:
is it possible after.get() got a null?

is it possible before.get() got a null?
if yes, then originalConfigMap must not be ConcurrentHashMap, then how to resolve my last question about "safe"

Reply:
Use ConcurrentHashMap instead.